fix: Update provisioning status when host has been deauthorized

cjnolan opened the pull request:
<!---
  SPDX-FileCopyrightText: (C) 2025 Intel Corporation
  SPDX-License-Identifier: Apache-2.0

  ------------------------------------------------------

  Author Mandatory (to be filled by PR Author/Submitter)
  ------------------------------------------------------

  - Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.
  - Those checklist items which are not marked are considered as not applicable for the PR change.
-->

### Description

Updates the onboarding manager to also set the provisioning status to failed when a host is deauthorized during provisioning. This ensures that the host status in the UI is set to the correct status in this case.

Fixes # (issue)

### Any Newly Introduced Dependencies

N/A

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

### Checklist:

- [x] I agree to use the APACHE-2.0 license for my code changes
- [x] I have not introduced any 3rd party dependency changes
- [x] I have performed a self-review of my code

pierventre commented on onboarding-manager/internal/handlers/controller/reconcilers/host.go:
```diff
@@ -51,6 +52,11 @@ func (hr *HostReconciler) Reconcile(ctx context.Context,
 		return directive
 	}
 
+	instance, err := hr.invClient.GetInstanceResourceByResourceID(ctx, tenantID, resourceID)
```
you dont need this. Host eager loads instance.

cjnolan commented:
> this approach is not going to work. We have suggested a different approach to solve this issue. do you see any problem in the proposed solution?

The Reconcile function updates the host status when a host gets deauthorized in the UI, but it does not update the other statuses in the instance, which causes the incorrect status to be reported in the UI. If the other statuses are updated there, based on the host state being changed, wouldn't that also set the correct overall status for the host when reported in the UI? Is there somewhere else that would then update the status again after the reconcile call, overwriting this?

The other approach suggested would work, but, since the reconcile updates then the host is deauthorized, would having it update all statuses also work as well?

pierventre commented:
Clean the status
- Host reconciler should clean also host related status. Double check that there are no other external stimula can lead to an update of the status
- Instance reconciler should look at the host state (unless UI does both Host/Instance de auth - unlikely looking at the ticket) and clean any related Instance status, delete the workflow in Tinker CRDs, blocking any future update 

Do the same work for OM and LOC-A OM

pierventre commented:
> > this approach is not going to work. We have suggested a different approach to solve this issue. do you see any problem in the proposed solution?
> 
> The Reconcile function updates the host status when a host gets deauthorized in the UI, but it does not update the other statuses in the instance, which causes the incorrect status to be reported in the UI. If the other statuses are updated there, based on the host state being changed, wouldn't that also set the correct overall status for the host when reported in the UI? Is there somewhere else that would then update the status again after the reconcile call, overwriting this?
> 
> The other approach suggested would work, but, since the reconcile updates then the host is deauthorized, would having it update all statuses also work as well?

Instance reconciler periodically reconciles the Instance and would overwrite the change implemented here

cjnolan commented:
> Instance reconciler periodically reconciles the Instance and would overwrite the change implemented here. Look at this issue from reconciler povs, external stimula and make sure to prevent continuos updates

Ok, I've switched the change to run as part of the instance reconciler instead, since that handles all of the statuses in the instance

pierventre reviewed (commented):
1) We are stilling missing host related statuses cleaning (iirc we update only host status but we should purge the other host modern status)

If we want to stick to what we discussed during the sync call:

2) We should not force Instance deauth - the user must do that. Irrespectively of that deauthorizing an host implicitly will make each component not functional. We should purge the content of each status and clean the provisioning resources
3) we should check if the status are purged already -> skip
4) we should check if the provisioning resources do not exist -> skip

We need to create end-states - so even if the reconciler re-run will be ineffective and not performing continuos writes

Now - I dont dislike this solution (explicitly de-auth the instance) too but there are still things missing like host handling, setting instance status to deauth (and not setting to failed), purging the other instance modern statuses

cjnolan commented:
> We are stilling missing host related statuses cleaning (iirc we update only host status but we should purge the other host modern status)

I've added clearing the onboarding and registration states to the host reconciler stage when it invalidates the host during deauth.

> We should not force Instance deauth - the user must do that. Irrespectively of that deauthorizing an host implicitly will make each component not functional. We should purge the content of each status and clean the provisioning resources

Removed.

> we should check if the status are purged already -> skip
we should check if the provisioning resources do not exist -> skip

Added to the instance reconciler, it will now, on detection of the host being in untrusted state, check the status indicators for instance, provisioning, update and trusted attestation to see if any are not in idle state and will update the statuses in that case. It will also check if a provisioning workflow exists for this instance and removes it. If both the status and workflow cleanup have been done in previous reconciliation loops, it will skip.

pierventre commented on onboarding-manager/internal/handlers/controller/reconcilers/host.go:
```diff
@@ -260,11 +260,17 @@ func (hr *HostReconciler) invalidateHost(ctx context.Context, host *computev1.Ho
+		CurrentState:                computev1.HostState_HOST_STATE_UNTRUSTED,
+		HostStatus:                  om_status.AuthorizationStatusInvalidated.Status,
+		HostStatusIndicator:         om_status.AuthorizationStatusInvalidated.StatusIndicator, // #nosec G115
+		HostStatusTimestamp:         uint64(time.Now().Unix()),                                // #nosec G115
```
maybe for the ts - we dont really need. IIRC Inventory has an hook doing that for free on updates

cjnolan replied:
Removed

pierventre commented on onboarding-manager/internal/invclient/invclient.go:
```diff
@@ -407,6 +407,12 @@ func (c *OnboardingInventoryClient) UpdateHostStateAndRuntimeStatus(ctx context.
 		computev1.HostResourceFieldHostStatus,
 		computev1.HostResourceFieldHostStatusIndicator,
 		computev1.HostResourceFieldHostStatusTimestamp,
+		computev1.HostResourceFieldOnboardingStatus,
```
`UpdateHostStateAndRuntimeStatus` is this a shared function? We risk to reset some fields

cjnolan replied:
It is only called from the `invalidateHost` path in the host reconciler

pierventre commented on onboarding-manager/internal/handlers/controller/reconcilers/instance.go:
```diff
@@ -85,9 +86,70 @@ func (ir *InstanceReconciler) Reconcile(ctx context.Context,
+		instance.GetHost().GetDesiredState() == computev1.HostState_HOST_STATE_UNTRUSTED {
+		checkCleanupRun := false
+		// Check that all statuses and indicators have been updated
+		if !checkStatusIdle(instance) {
```
at least for this part of the code shall we add a function `checkStatusEmpty` ?

pierventre commented on onboarding-manager/internal/handlers/controller/reconcilers/instance.go:
```diff
@@ -85,9 +86,70 @@ func (ir *InstanceReconciler) Reconcile(ctx context.Context,
+			// Run reconciliation on the instance to complete
+			zlogInst.Info().Msgf("Host associated with Instance (%s) has been deauthorized. "+
+				"Forcing reconciliation to update Instance status.", resourceID)
+			return ir.reconcileInstance(ctx, request, instance)
```
use requeue or resubmti directive?

cjnolan replied:
Since we have removed the deauth of the instance from this, we can return `request.Ack()` instead after updating the statuses and removing the workflows

cjnolan commented on onboarding-manager/internal/handlers/controller/reconcilers/instance.go:
```diff
@@ -88,6 +93,71 @@ func (ir *InstanceReconciler) Reconcile(ctx context.Context,
+					return request.Ack()
+				}
+			}
+			ir.updateInstanceStatuses(ctx, instance)
```
It is needed to update the inventory with the new instance statues. If this not done, the instance is not updated and the next loop through the reeconciler will have the same statuses

pierventre reviewed (approved):
@Ram-srini @ipsita-npg @ppanigra please review

cjnolan commented on onboarding-manager/internal/handlers/controller/reconcilers/instance.go:
```diff
@@ -153,6 +223,28 @@ func (ir *InstanceReconciler) handleMatchingStates(ctx context.Context, instance
+func (ir *InstanceReconciler) updateInstanceStatuses(
+	ctx context.Context,
+	newInstance *computev1.InstanceResource,
+) {
```
Done

pierventre reviewed (approved):
love it
